client/src/lib/utils.ts

```diff
@@ -9,3 +12,24 @@ export function cn(...inputs: ClassValue[]) {
 export function isEmptyWidget(data: BaseWidgetWithData["data"]): boolean {
   return Object.keys(data).length === 0;
 }
+
+export function normalizeWidgetData(widgetData: WidgetData): WidgetData {
```

Comment:
As suggestion, I would rather prefer going with principle of single responsibility and split this function into two different ones: one to handle data for the map and other for data for charts.

It might look like simple and easy to understand now, but if we start adding more layers of complexity to the parsing for any reason, it will also harder to understand and probably prone to errors.

Said that, it's just a suggestion, up to you!